src/flag_gems/utils/libentry.py

```diff
@@ -151,13 +156,127 @@ def decorator(fn):
     return decorator
 
 
+def register_should_search_hook(name):
+    def decorator(fn):
+        SHOULD_SEARCH_HOOKS[name] = fn
+        return fn
+
+    return decorator
+
+
+@register_should_search_hook("default")
+def default_should_search_hook(tuner, key):
+    return key not in tuner.cache
+
+
+@register_should_search_hook("bayesian")
+def bayesian_should_search_hook(tuner, key):
+    if key not in tuner.cache:
+        return True
+    tuner.call_counts[key] = tuner.call_counts.get(key, 0) + 1
+    # Starts with a 20% chance of exploration, decays over calls
+    explore_prob = 0.2 * (0.95 ** (tuner.call_counts[key] - 1))
+    return random.random() < explore_prob
+
+
 @register_search_strategy("brute")
-def default_search_strategy(bench_fn, configs, args, kwargs):
+def brute_search_strategy(bench_fn, configs, args, kwargs, tuner, key):
     timings = {config: bench_fn(config) for config in configs}
     best_config = builtins.min(timings, key=timings.get)
     return best_config, timings
 
 
+@register_search_strategy("early_stop")
+def early_stop_search_strategy(bench_fn, configs, args, kwargs, tuner, key):
+    timings = {}
+    best_config = None
+    best_timing = float("inf")
+    second_best_timing = float("inf")
+    # Stop if the relative difference between the best and second-best is less than x%
+    # Only start checking after this many configs
+    threshold = 0.03
+    min_search_before_stop = 5
+    search_count = 0
+
+    for config in configs:
+        search_count += 1
+        timing = bench_fn(config)[0]
+        timings[config] = timing
+
+        if timing < best_timing:
+            second_best_timing = best_timing
+            best_timing = timing
+            best_config = config
+        elif timing < second_best_timing:
+            second_best_timing = timing
+
+        # Check for early stopping condition
+        if (
+            search_count >= min_search_before_stop
+            and (second_best_timing - best_timing) / best_timing < threshold
+        ):
+            break
+    return best_config, timings
+
+
+@register_search_strategy("bayesian")
+def bayesian_search_strategy(bench_fn, configs, args, kwargs, tuner, key):
+    timings = tuner.timings_cache.get(key, {})
+
+    # Define search space
+    search_space = []
+    if not configs:
+        return None, {}
+
+    param_names = list(configs[0].kwargs.keys())
+    for name in param_names:
+        values = sorted(list(set(c.kwargs[name] for c in configs)))
```

Comment:
Looks like you don't need to convert it into `list`.
```suggestion
        values = sorted(set(c.kwargs[name] for c in configs))
```